WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Package Selection/Package Creation/ShippingLabelAddNewPackage.swift

```diff
@@ -2,8 +2,7 @@ import SwiftUI
 import Yosemite
 
 struct ShippingLabelAddNewPackage: View {
-    @StateObject private var viewModel = ShippingLabelAddNewPackageViewModel()
-    let packagesResponse: ShippingLabelPackagesResponse?
+    @StateObject var viewModel: ShippingLabelAddNewPackageViewModel
```

Comment:
I'm not quite comfortable with using `StateObject` here, although it helps persisting the state of the the object. I've found [this article](https://www.donnywals.com/whats-the-difference-between-stateobject-and-observedobject/), and I'm interested in this sentence:

> If you'd have marked Counter.dataSource as @StateObject the new instance would be ignored and your app might now have a new hidden bug.

So the takeaway is - like @State - @StateObject should only be used if the object is created and owned by the view itself. @State and @StateObject should only be used to persist the state when the view is **redrawn**, not when it's **created**.

In our case, when rotating the device, `ShippingLabelPackageList` view is redrawn, which causes `ShippingLabelAddNewPackage` to be recreated. So we want to persist `ShippingLabelAddNewPackage`'s last state even after it's created, and we should make `ShippingLabelPackageDetailsViewModel` hold on to its state like so:
```Swift
final class ShippingLabelPackageDetailsViewModel: ObservableObject {
    lazy var addNewViewModel: ShippingLabelAddNewPackageViewModel = ShippingLabelAddNewPackageViewModel(packagesResponse: packagesResponse)
}
```

And we use this `addNewViewModel` when creating the `ShippingLabelAddNewPackage` view instead of creating a new view model every time:
```Swift
NavigationLink(destination: ShippingLabelAddNewPackage(viewModel: viewModel.addNewViewModel),
               isActive: $isShowingNewPackageCreation) {
    EmptyView()
}
```

We can then make the view model an @ObservedObject in `ShippingLabelAddNewPackage` and still keep its state when it's created again. That way we delegate the view's state persistance to its parent view, while avoiding incorrect use of @StateObject. What do you think - does that make sense?

Reply:
Thanks for this explanation, @itsmeichigo!

> So the takeaway is - like @State - @StateObject should only be used if the object is created and owned by the view itself. @State and @StateObject should only be used to persist the state when the view is redrawn, not when it's created.

I remember reading about this earlier but I had lost sight of it in the process of trying to get everything to work with these views. Thanks for the reminder! With that mind then definitely it makes sense to use `ShippingLabelPackageDetailsViewModel` to hold the view model for `ShippingLabelAddNewPackage`. I'll work on that and take a look through to make sure I'm not misusing `StateObject` anywhere else.